feat: generate zsh shell completion script during build process using…

Rhahkeem opened the pull request:
… clap_complete

This generates the completions for the binary invocation during the compilation process. This ensures that as long as the binary can build that zsh will be able to use the autocomplete for the different commands and sub commands that exist.

Commit:
feat: generate zsh shell completion script during build process using clap_complete

This generates the completions for the binary invocation during the compilation process.
This ensures that as long as the binary can build that zsh will be able to use the
autocomplete for the different commands and sub commands that exist.

Commit:
Changed after feedback

craftytrickster reviewed (commented):
Looks mostly fine.

Not sure if you are aware of this, but it is general better in my opinion to put the help for clap structs directly on the fields as doc strings.
```
/// any description you give over a field like this will automatically
/// become the help in the help attribute of the macro
```
`    #[clap(help = "foobar")]`

This has the added benefit of also giving docstring hints to the actual code itself when being used by an IDE

Commit:
Reformat and cleanup

This commit cleans up deprecated API's and replaces some of the `ensure`
calls with a suitable `match` replacement.

Commit:
Add docstrings for argument structs

This adds docstrings to all the command structs so that helps with
generation of documentation.

nolram88 reviewed (approved):
Looks good to me. The change set is focused, the coverage story is solid, and I did not spot any obvious issues in the touched paths.
